### PR TITLE
debug handler: add /debug/request http handlers

### PR DESCRIPTION
**Pull request by mwitkow**
This adds `/debug/request` handling for the server and client side. 
Fixes issues with empty transport of `DefaultTransport`. 
Also fixes error propagation with opentracing by adding a test.

### CODE REVIEW

File: tracing/opentracing/integration_test.go [hunk 1]
```diff
@@ -135,3 +135,13 @@ func (s *OpentracingSuite) TestPropagatesErrors() {
...
 	assert.EqualValues(s.T(), 501, serverSpan.Tag("http.status_code"), "server span needs the correct status code marking")
 }
+
+func (s *OpentracingSuite) TestTripperwareHandlesErrors() {
+	client := s.WaresTestSuite.ClientTripperware.WrapClient(http.DefaultClient)
+	ctx := s.createContextFromFakeHttpRequestParent(s.SimpleCtx())
+	req, _ := http.NewRequest("POST", "https://whatever.doesntexist/someurl?code=501", nil)
+	req = http_ctxtags.TagRequest(req.WithContext(ctx), "assert_service", "assert_method")
+	_, err := client.Do(req)
+	require.Error(s.T(), err, "call shouldn't fail")
```
Comment (improbable-yifan) [inline tracing/opentracing/integration_test.go]: update the comment to "call should fail"

File: tracing/debug/options.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// FilterFunc allows users to provide a function that filters out certain methods from being traced.
+//
+// If it returns false, the given request will not be traced.
+type FilterFunc func(req *http.Request) bool
+
+// StatusCodeIsError allows the customization of which requests are considered errors in the tracing system.
+type StatusCodeIsError func(statusCode int) bool
+
+type options struct {
+	filterOutFunc       FilterFunc
```
Comment (improbable-yifan) [inline tracing/debug/options.go]: `s/filterOutFunc/filterFunc/`

File: tracing/debug/options.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	}
+)
+
+// FilterFunc allows users to provide a function that filters out certain methods from being traced.
+//
+// If it returns false, the given request will not be traced.
+type FilterFunc func(req *http.Request) bool
+
+// StatusCodeIsError allows the customization of which requests are considered errors in the tracing system.
+type StatusCodeIsError func(statusCode int) bool
```
Comment (improbable-yifan) [inline tracing/debug/options.go]: `IsStatusCodeError`?

File: tracing/debug/options.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		o.statusCodeErrorFunc = f
+	}
+}
+
+// DefaultStatusCodeIsError defines a function that says whether a given request is an error based on a code.
+func DefaultStatusCodeIsError(statusCode int) bool {
+	if statusCode < 500 {
+		return false
+	}
+	return true
```
Comment (improbable-yifan) [inline tracing/debug/options.go]: just `return statusCode >= 500`

File: tracing/debug/tripperware.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/mwitkow/go-httpwares"
+	"github.com/mwitkow/go-httpwares/tags"
+	"golang.org/x/net/trace"
+)
+
+// Tripperware returns a piece of client-side Tripperware that puts requests on a status page.
```
Comment (improbable-yifan) [inline tracing/debug/tripperware.go]: worth pointing out what is logged to the status page

File: tracing/debug/tripperware.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+				tr.LazyPrintf("Response: %d, length: %d", resp.Status, resp.ContentLength)
+				for k, _ := range resp.Header {
+					tr.LazyPrintf("%v: %v", k, resp.Header.Get(k))
+				}
+				if o.statusCodeErrorFunc(resp.StatusCode) {
+					tr.SetError()
+				}
+			}
+			tr.Finish()
+			return resp, err
```
Comment (improbable-yifan) [inline tracing/debug/tripperware.go]: put this next to the trace creation with a defer.

File: tracing/opentracing/integration_test.go (hunk 1 above)
Comment (mwitkow) [inline tracing/opentracing/integration_test.go]: Done.

File: tracing/debug/options.go (hunk 2 above)
Comment (mwitkow) [inline tracing/debug/options.go]: done

File: tracing/debug/options.go (hunk 4 above)
Comment (mwitkow) [inline tracing/debug/options.go]: done

File: tracing/debug/tripperware.go (hunk 5 above)
Comment (mwitkow) [inline tracing/debug/tripperware.go]: done

File: tracing/debug/tripperware.go (hunk 6 above)
Comment (mwitkow) [inline tracing/debug/tripperware.go]: done.

**Comment by mwitkow** [comment]
Addressed @yifanzz comments.

**Pull request merged**